lib/SPIRVProducerPass.cpp

```diff
@@ -3220,191 +3283,850 @@ spv::Op SPIRVProducerPass::GetSPIRVBinaryOpcode(Instruction &I) {
+
+    // May require an extra instruction to create the appropriate result of
+    // the builtin function.
+    if (FuncInfo.getType() == Builtins::kGetImageDim) {
+      if (dim == 3) {
+        // get_image_dim returns an int4 for 3D images.
+        //
+
+        // Implement:
+        //   %result = OpCompositeConstruct %uint4 %sizes %uint_0
+        SPIRVID CstInt0 = getSPIRVInt32Constant(0);
+
+        Ops.clear();
+        Ops << FixedVectorType::get(Type::getInt32Ty(Context), 4) << RID
+            << CstInt0;
+
+        RID = addSPIRVInst(spv::OpCompositeConstruct, Ops);
+      } else if (dim != components) {
+        // get_image_dim return an int2 regardless of the arrayedness of the
+        // image. If the image is arrayed an element must be dropped from the
+        // query result.
+        //
+
+        // Implement:
+        //   %result = OpVectorShuffle %uint2 %sizes %sizes 0 1
+        Ops.clear();
+        Ops << FixedVectorType::get(Type::getInt32Ty(Context), 2) << RID << RID
+            << 0 << 1;
+
+        RID = addSPIRVInst(spv::OpVectorShuffle, Ops);
+      }
+    } else if (components > 1) {
+      // Implement:
+      //     %result = OpCompositeExtract %uint %sizes <component number>
+      Ops.clear();
+      Ops << Call->getType() << RID;
+
+      uint32_t component = 0;
+      if (FuncInfo.getType() == Builtins::kGetImageHeight)
+        component = 1;
+      else if (FuncInfo.getType() == Builtins::kGetImageDepth)
+        component = 2;
+      Ops << component;
+
+      RID = addSPIRVInst(spv::OpCompositeExtract, Ops);
+    }
+    break;
+  }
+  default:
+    llvm_unreachable("Unsupported Image builtin");
+  }
+
+  return RID;
+}
+
+SPIRVID
+SPIRVProducerPass::GenerateSubgroupInstruction(CallInst *Call,
+                                               const FunctionInfo &FuncInfo) {
+  SPIRVID RID;
+
+  // requires SPIRV version 1.3 or greater
+  if (SPIRVVersion < Version(1, 3)) {
+    // llvm_unreachable("SubGroups extension requires SPIRV 1.3 or greater");
+    // for now just upgrade the version
+    SPIRVVersion.setMinor(3);
+  }
+
+  auto loadBuiltin = [this, Call](spv::BuiltIn spvBI,
+                                  spv::Capability spvCap =
+                                      spv::CapabilityGroupNonUniform) {
+    SPIRVOperandVec Ops;
+    Ops << Call->getType() << this->getSPIRVBuiltin(spvBI, spvCap);
+
+    return addSPIRVInst(spv::OpLoad, Ops);
+  };
+
+  spv::Op op = spv::OpNop;
+  switch (FuncInfo.getType()) {
+  case Builtins::kGetSubGroupSize:
+    return loadBuiltin(spv::BuiltInSubgroupSize);
+  case Builtins::kGetNumSubGroups:
+    return loadBuiltin(spv::BuiltInNumSubgroups);
+  case Builtins::kGetSubGroupId:
+    return loadBuiltin(spv::BuiltInSubgroupId);
+  case Builtins::kGetSubGroupLocalId:
+    return loadBuiltin(spv::BuiltInSubgroupLocalInvocationId);
+  case Builtins::kGetEnqueuedNumSubGroups:
+    return loadBuiltin(spv::BuiltInNumEnqueuedSubgroups, spv::CapabilityKernel);
```

Comment:
Kernel is incompatible with Shader. We can't implement num_enqueued_subgroups this way. Leave it unimplemented. Update the docs too.